Add support for Open SSL 3.0

The latest version of OpenSSL 3.0 removes some api calls such as EVP_PKEY_size.

Compilation simply fails

```
/usr/bin/ld: O-penS-S-L-5858P-K-ey5858R-S-A-.o: in function `*OpenSSL::PKey::RSA@OpenSSL::PKey::PKey#max_encrypt_size:Int32':
OpenSSL::PKey::RSA:(.text+0x756): undefined reference to `EVP_PKEY_size'
collect2: error: ld returned 1 exit status
```

I wonder if there is a way to make the code forward compatible with Openssl 3.0.

I just ran in to this too from using the JWT shard.

```
web          | /usr/bin/ld: O-penS-S-L-5858P-K-ey5858R-S-A-.o: in function `max_encrypt_size':
web          | /home/jeremy/Sites/joysticktv/lib/openssl_ext/src/openssl_ext/pkey/pkey.cr:242: undefined reference to `EVP_PKEY_size'
web          | collect2: error: ld returned 1 exit status
web          | Error: execution of command failed with code: 1: `cc "${@}" -o /home/jeremy/Sites/joysticktv/bin/start_server  -rdynamic -L/usr/bin/../lib/crystal -lxml2 -lgmp -lyaml `command -v pkg-config > /dev/null && pkg-config --libs --silence-errors libssl || printf %s '-lssl -lcrypto'` `command -v pkg-config > /dev/null && pkg-config --libs --silence-errors libcrypto || printf %s '-lcrypto'` -lz -lpcre -lm -lgc -lpthread -levent -lrt -lpthread -ldl`
```

--- a/src/openssl_ext/lib_crypto.cr
+++ b/src/openssl_ext/lib_crypto.cr
@@ -264,10 +264,21 @@ lib LibCrypto
   fun bio_ctrl = BIO_ctrl(bio : Bio*, cmd : LibC::Int, larg : LibC::Long, parg : Void*) : LibC::Long
 
   fun evp_pkey_new = EVP_PKEY_new : EvpPKey*
-  fun evp_pkey_id = EVP_PKEY_id(pkey : EvpPKey*) : LibC::Int
   fun evp_pkey_free = EVP_PKEY_free(pkey : EvpPKey*)
-  fun evp_pkey_size = EVP_PKEY_size(pkey : EvpPKey*) : LibC::Int
-  fun evp_pkey_bits = EVP_PKEY_bits(pkey : EvpPKey*) : LibC::Int
+  {% if compare_versions(LibCrypto::OPENSSL_VERSION, "3.0.0") >= 0 %}
+    fun evp_pkey_id = EVP_PKEY_get_id(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_base_id = EVP_PKEY_get_base_id(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_size = EVP_PKEY_get_size(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_bits = EVP_PKEY_get_bits(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_security_bits = EVP_PKEY_get_security_bits(pkey : EvpPKey*) : LibC::Int
+  {% else %}
+    fun evp_pkey_id = EVP_PKEY_id(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_base_id = EVP_PKEY_base_id(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_size = EVP_PKEY_size(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_bits = EVP_PKEY_bits(pkey : EvpPKey*) : LibC::Int
+    fun evp_pkey_security_bits = EVP_PKEY_security_bits(pkey : EvpPKey*) : LibC::Int
+  {% end %}
+
   fun evp_pkey_get0 = EVP_PKEY_get0(pkey : EvpPKey*) : Void*
   fun evp_pkey_get0_rsa = EVP_PKEY_get0_RSA(pkey : EvpPKey*) : Rsa*
   fun evp_pkey_get0_dsa = EVP_PKEY_get0_DSA(pkey : EvpPKey*) : Dsa*

